### ISSUE TITLE
Precondition errors when using ArrayList add(o)

### ISSUE THREAD

**Issue by ledru** [OpenJML/OpenJML#486]
Hi,

I have problems when compiling a plain java program, which uses ArrayLists, with openJML.
I am using OpenJML 0.8.6 , eclipse Neon 2, windows 10, and java 8.

Here is the program:

```
import java.util.ArrayList;

public class EssaiArrayList {
	
	public static ArrayList<Integer> ali = new ArrayList<Integer>();
	
	public static void main(String[] args) {
		ali.add(1);
	}
}
```

If I compile this program with openjml rac, and then execute it, 
I get a precondition error related to the add() method.

Here is the error I get in eclipse Neon2 (with options Skip purity check, Check assumptions, Distinguish entry preconditions):
```

Exception in thread "main" org.jmlspecs.utils.JmlAssertionError$Precondition: C:\Users\ledru\workspaces\Neon2_JML\Problems\src\EssaiArrayList.java:8: JML precondition is false
		ali.add(1);
		       ^
C:\EclipseNeon2JML\configuration\org.eclipse.osgi\471\0\.cp\java5\java\util\Collection.jml:182: Associated declaration: C:\Users\ledru\workspaces\Neon2_JML\Problems\src\EssaiArrayList.java:8: 
      @   requires !containsNull ==> o != null;
          ^
	at org.jmlspecs.utils.Utils.convertPrecondition(Utils.java:103)
	at EssaiArrayList.main(EssaiArrayList.java:7)

```
	
I have a similar error message when using command-line :

```
>java -jar "..\..\OpenJML\openjml.jar"  -rac -nopuritycheck -racCheckAssumptions -racPreconditionEntry -d ..\bin EssaiArrayList.java

>java -cp ..\bin;.;"..\..\OpenJML\jmlruntime.jar";"..\..\OpenJML\jmlspecs.jar" EssaiArrayList
Skipping a specification clause because it contains an uncompiled model method: java.util.ArrayList.initialArrayList
Skipping a specification clause because it contains an uncompiled ghost or model field: containsNull
Skipping a specification clause because it contains an uncompiled ghost or model field: containsNull
Skipping a specification clause because it contains an uncompiled ghost or model field: containsNull
Skipping a specification clause because it contains an uncompiled ghost or model field: containsNull
Skipping a specification clause because it contains an uncompiled ghost or model field: containsNull
Skipping a specification clause because it contains an uncompiled ghost or model field: containsNull
Skipping a specification clause because it contains an uncompiled ghost or model field: containsNull
Skipping a specification clause because it contains an uncompiled ghost or model field: containsNull
Exception in thread "main" org.jmlspecs.utils.JmlAssertionError$Precondition: EssaiArrayList.java:8: JML precondition is false
                ali.add(1);
                       ^
..\..\OpenJML\openjml.jar(specs18/java/util/Collection.jml):182: Associated declaration: EssaiArrayList.java:8:
      @   requires !containsNull ==> o != null;
          ^
        at org.jmlspecs.utils.Utils.convertPrecondition(Utils.java:103)
        at EssaiArrayList.main(EssaiArrayList.java:7)

```

Thanks by advance for your help.

Yves

**Comment by davidcok** [comment]
The issue here is that Collections.jml has some specifications (for add) that are not executable when the standard library jar is used. In fact all of the behaviors of add are in this category and so all preconditions default to false -- consequently at runtime a complaint is issued that the composite precondition is false.

The fix is to edit those preconditions in the specs so that they are not used for RAC. In that case this test case behaves as desired. For usability, I'm also adding location information about the skipped specification clauses and some other usability helps.

This is likely the case for other method specs as well, but I have not yet reviewed all the library specs to avoid this problem.

When full testing is completed, these changes will be committed. Do you need an interim build to be able to continue working?

**Comment by ledru** [comment]
Thank you David!

> Do you need an interim build to be able to continue working?

I will give some homework to my students on Feb. 6th, and some of them may use collections. So I'd be glad to have a build (interim or full) by that date. Thanks by advance!

Yves

**Comment by davidcok** [comment]
I will make a build next week in time for your class. However, the specifications for collections when used by RAC need some work - as you discovered - I will do some fixing before I make a release, but it would be helpful to know a reasonable set of classes you expect the students to use, particularly collections classes.
- David

**Comment by ledru** [comment]
> I will make a build next week in time for your class. However, the specifications for collections when used by RAC need some work - as you discovered - I will do some fixing before I make a release, but it would be helpful to know a reasonable set of classes you expect the students to use, particularly collections classes. 

Last year, the students used ArrayList, HashSet and Set. The priority is with ArrayList.

Thanks by advance

Yves

**Comment by MikeRawding** [comment]
Hi David,

Will you be creating a branch to handle this issue?  Perhaps I can provide some assistance by extending this work to other collections.

**Comment by davidcok** [comment]
I will be making a release this week.

**Comment by davidcok** [comment]
The specifications for RAC for ArrayList, Set, HashSet, List, Collection have been improved and released in V0.8.7. A thorough review of specs for Collection classes from a RAC point of view is still needed.
